Enhanced ResourceTemplate URI matching and type handling logic.

beaterblank opened the pull request:
<!-- Provide a brief summary of your changes -->

**Feature Added:**
Enhanced `ResourceTemplate` URI matching and type handling logic.

## Motivation and Context

This change introduces a typed-aware URI parsing mechanism similar to FastAPI’s path parameter system.
It resolves issue **#220** by adding a more flexible and self-contained approach for handling resource URIs with type annotations.

## How Has This Been Tested?

* Verified functionality through new and existing unit tests for:

  * Typed placeholders (`{name:str}`, `{id:int}`, `{value:float}`, `{uuid:UUID}`, `{path:path}`)
  * Automatic type conversion and validation
  
* Confirmed correct parameter extraction and type enforcement in all matching scenarios.

## Breaking Changes

* None expected, but users relying on untyped string-only behaviour may observe typed conversions.

## Types of changes

* [x] New feature (non-breaking change which adds functionality)


## Checklist

* [x] I have read the [[MCP Documentation](https://modelcontextprotocol.io/)](https://modelcontextprotocol.io)
* [x] My code follows the repository's style guidelines
* [x] New and existing tests pass locally
* [x] I have added appropriate error handling
* [x] I have added or updated documentation as needed

## Additional context

The implementation assumes URI templates follow a strict pattern and defaults to `str` when no type is provided.
Optional parameters and wildcard matching beyond `{path: path}` are not supported / cannot be supported.

Commit:
Added RFC 6570 complaint form style query expansion as optional parameters to resources

Commit:
uv lock fix

Commit:
fix

Commit:
fix mismatch

Commit:
added type validation for optional params

Commit:
resolve conflicts

Commit:
fix format

Commit:
resolved conflicts

Commit:
resolved conflicts

beaterblank committed:
Feature Added:
Improve ResourceTemplate URI matching and type handling

Fixes issue #220

- Implemented typed-aware URI matching logic within ResourceTemplate.matches()
  supporting {name:str}, {id:int}, {value:float},{uuid:UUID}, and {path:path} placeholders.
  similar to what FastAPI supports
- Added automatic type conversion and validation for URI parameters.
- Updated unit tests to cover typed placeholders, numeric and float parameters,
  and path-based URIs.

Limitations and notes:
- Current implementation assumes URI templates follow a strict pattern or default
  to str and do not support optional parameters or wildcards beyond {path:path}.

beaterblank committed:
linting and formatting

beaterblank committed:
Enhancement: compile and store the pattern once instead of recompiling it on every match, removed bare exception added value error

beaterblank committed:
formatting

beaterblank committed:
Merge branch 'main' into main

beaterblank committed:
Merge branch 'main' into main

beaterblank committed:
Merge branch 'main' into main

felixweinberger reviewed (commented):
Hi @beaterblank thank you for this contribution - FYI there's another PR https://github.com/modelcontextprotocol/python-sdk/pull/427 that's also enhancing URI handling, though for a slightly different usecase, might be some conflicts.

This PR takes the convertor approach suggested in #220, but @Kludex raised a good point on #427 whether we should aim for more of a FastAPI-level design, which makes sense in terms of abstraction level to me.

Rather than embedding convertor syntax in the URI template `({path:path}, {id:int})`, we could specify parameter behavior in the function signature using Annotated:

```python
from typing import Annotated

@mcp.resource("files://{path}")
def get_file(path: Annotated[str, Path(allow_slashes=True)]):
    ...
```

This would also provide a unified way to handle both path and query parameters:

```python
@mcp.resource("search://{category}")
def search(
    category: str,  # Simple path param, no annotation needed
    format: Annotated[str, Query(default="json")],  # Query param
    limit: Annotated[int, Query(default=10)]
):
    ...
```

beaterblank commented on src/mcp/server/fastmcp/resources/templates.py:
```diff
@@ -79,14 +82,54 @@ def from_function(
+            try:
+                result[name] = conv.convert(raw_value)
+            except Exception as e:
+                raise ValueError(f"Failed to convert '{raw_value}' for '{name}': {e}")
```
Should This be a `ValueError`? Since it only happens when the converter has a bad implementation of converting.

Should we go for `RuntimeError`?

beaterblank commented:
> > Hi @beaterblank thank you for this contribution - FYI there's another PR #427 that's also enhancing URI handling, though for a slightly different usecase, might be some conflicts.
> > This PR takes the convertor approach suggested in #220, but @Kludex raised a good point on #427 whether we should aim for more of a FastAPI-level design, which makes sense in terms of abstraction level to me.
> > Rather than embedding convertor syntax in the URI template `({path:path}, {id:int})`, we could specify parameter behavior in the function signature using Annotated:
> > ```python
> > from typing import Annotated
> > 
> > @mcp.resource("files://{path}")
> > def get_file(path: Annotated[str, Path(allow_slashes=True)]):
> >     ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This would also provide a unified way to handle both path and query parameters:
> > ```python
> > @mcp.resource("search://{category}")
> > def search(
> >     category: str,  # Simple path param, no annotation needed
> >     format: Annotated[str, Query(default="json")],  # Query param
> >     limit: Annotated[int, Query(default=10)]
> > ):
> >     ...
> > ```
> 
> Yes, I've also just noticed this, and this is partly done in #427 ,

Merging #427 into this.

beaterblank committed:
Merge branch 'main' into main

beaterblank committed:
merge #427

beaterblank committed:
handle validation in templates

beaterblank committed:
bug fix

beaterblank committed:
linting

beaterblank committed:
Merge branch 'main' into main

beaterblank commented:
Merged #427 

Currently working on the Annotated parameter logic implementation, and addressing comments from #427.
Reference: [fastapi/param_functions.py](https://github.com/fastapi/fastapi/blob/master/fastapi/param_functions.py)

While reviewing the merge, I noticed this line in the MCP SDK:
[https://github.com/modelcontextprotocol/python-sdk/blob/40acbc596c727e8ccc58c69df8e93e9cf6e56a4e/src/mcp/server/fastmcp/server.py#L1073](https://github.com/modelcontextprotocol/python-sdk/blob/40acbc596c727e8ccc58c69df8e93e9cf6e56a4e/src/mcp/server/fastmcp/server.py#L1073)

Should we leave this as-is and handle **`Path`** and **`Query`** separately, or try to integrate context injection here as well?
Integrating it could introduce regressions or break existing behaviour.

beaterblank committed:
added Path and Query Annotations

beaterblank committed:
Merge branch 'main' of https://github.com/beaterblank/python-sdk

beaterblank commented:
The code ended up being longer than expected since much of it is boilerplate. I made as close to FastAPI’s implementation as I could, omitting any deprecated components. I can provide a diff or list of differences if needed, but they should be easy to reintroduce later if necessary.

I also had to jump loops to handle the `Context` parameter; the current approach should be the better way to implement it. I avoided further edits in that area to keep this PR from growing even larger.

@felixweinberger  Please review the code. Thank you!

beaterblank committed:
linting

beaterblank commented:
More tests, examples, and README updates should be added for this. It introduces a major non-breaking change.

beaterblank committed:
examples

beaterblank committed:
Path and Query bug fixes and tests and example.

beaterblank committed:
Merge branch 'main' into main

beaterblank committed:
Merge branch 'main' into main

beaterblank committed:
documentation corrections

beaterblank committed:
test case fixes

BossChaos commented:
## Code Review

**PR:** [Enhanced ResourceTemplate URI matching and type handling logic.](https://github.com/modelcontextprotocol/python-sdk/pull/1439) by @beaterblank

- ✅ Code review

**Wallet:** `0xdaE5d307339074A24F579dB48e7c639359D94904`

*Code review*

maxisbey commented:
Thanks for the PR. This has since landed via #2356. Closing as part of a general backlog cleanup following the v2 release. If this is still relevant against v2, feel free to reopen.

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>
